### PR TITLE
Stream management page

### PR DESCRIPTION
**Pull request by kjk2010**
Here's a simple page that uses the API to quickly add/edit/delete streams from an AmpliPi. Not sure if it's of any interest to anyone, but wanted to offer it in case it was useful.

### CODE REVIEW

**Comment by linknum23** [comment]
I just showed this to everyone in the office. All of us are very impressed at how good it looks and how it matches the rest of the app's look and feel.

We are going to try and get this integrated before we ship the early bird units since it will simplify basic interactions with AmpliPi. 

Below are some changes that we are planning to make before integrating this:

- [ ] Integrate into settings page (adding a drop down menu for audio inputs)
- [ ] Instead of `()`, colorize the stream type [blue]() to differentiate type from name
- [ ] Integrate plex auth sequence (plex needs special credentials)
- [ ] Add stream info validation during creation (on web side handle failures with a message, server side needs to test playback to /dev/null and verify no errors)
- [ ] Integrate analog audio sources (for name change, not delete-able)
- [ ] Add note somewhere that separate streams are needed for simultaneous playback (eg. 2 Spotify Streams)

I am going to make this a draft PR until most of the above changes are finished. Let us know if you would like to make any of the changes above, @micronova-jb and I will start looking into the best way to integrate all of this.

This is really going to make it simpler for everyone to interface with AmpliPi, thanks!

**Comment by kjk2010** [comment]
I'd be glad to work on at least some of those items. I probably won't be able to push an update until some time next week though.

One item I'd like to add to the list is the ability to re-order the stream list, possibly by adding an 'order' value to each of the streams. It would allow for ordering by importance, or even alphabetically.

Regarding the second item on the list, will this work?
image

**Comment by linknum23** [comment]
That looks great. Please commit that change!

 I'll take a first pass at migrating this into the settings and then we can go from there.

For the ordering, I think we can add an **order** field that defaults to streamid, the current ordering method and is settable via the api after the stream has been created.

**Comment by kjk2010** [comment]
Change committed. Sounds good!

**Comment by linknum23** [comment]
I rebased this on master and make an initial pass at what the settings menu/nav should look like. The stream management is integrated into the settings page. Click **Gear icon** -> **Settings - Inputs** to navigate to the integrated stream management settings page.

See my comment above for what is left. I will try to keep the tasks in it up to date.

**Review by micronova-jb — APPROVED** [review]

File: web/static/js/script.js [hunk 1]
```diff
@@ -69,13 +69,14 @@ $(document).ready(function(){
       $('#' + old_src + '-input')[0].style.display = "none";
       $('#' + old_src + '-player')[0].style.display = "none";
     } else {
-      $('#settings-header')[0].style.display = "none";
+      var src = new_src[1];
```
Comment (linknum23) [inline web/static/js/script.js:72]:
```suggestion
      // reload the page, to keep it up to date with any changes made in settings   
      var src = new_src[1];
```

File: web/static/js/stream-management.js [hunk 2]
```diff
@@ -0,0 +1,488 @@
...
+  $.get("/api/streams", function(data) {
+    $.each(data.streams, function(k, v) {
+      streams[v.id] = v;
+      $("#settings-tab-inputs-stream-selection").append(
+        '<li class="list-group-item list-group-item-action list-group-item-dark stream" style="vertical-align: bottom;" data-id="' + v.id + '">' +
+        v.name +
+        ' <span style="float:right;font-size:0.8rem;color:navy;line-height:25px;vertical-align: bottom;">' + v.type + '</span>'
+      );
+    });
+  });
```
Comment (linknum23) [inline web/static/js/stream-management.js:24]: It might make sense to combine the requests into one (eventually the zone and group mgmt code would probably be added here as well
```suggestion
  $.get("/api", function(data) {
    $.each(data.sources, function(k, v) {
      streams[v.id] = v;
      $("#settings-tab-inputs-stream-selection").append(
        '<li class="list-group-item list-group-item-action list-group-item-dark stream" style="vertical-align: bottom;" data-id="' + v.id + '">' +
        v.name +
        ' <span style="float:right;font-size:0.8rem;color:navy;line-height:25px;vertical-align: bottom;">' + `rca ${v.id+1}` + '</span>'
      );
    });
    $.each(data.streams, function(k, v) {
      streams[v.id] = v;
      $("#settings-tab-inputs-stream-selection").append(
        '<li class="list-group-item list-group-item-action list-group-item-dark stream" style="vertical-align: bottom;" data-id="' + v.id + '">' +
        v.name +
        ' <span style="float:right;font-size:0.8rem;color:navy;line-height:25px;vertical-align: bottom;">' + v.type + '</span>'
      );
    });
  });
```

File: web/static/js/stream-management.js [hunk 3]
```diff
@@ -0,0 +1,488 @@
...
+        $.get("/api/streams", function(data) {
+          $.each(data.streams, function(k, v) {
+            streams[v.id] = v;
+            $("#settings-tab-inputs-stream-selection").append(
+              '<li class="list-group-item list-group-item-action list-group-item-dark stream" style="vertical-align: bottom;" data-id="' + v.id + '">' +
+              v.name +
+              ' <span style="float:right;font-size:0.8rem;color:navy;line-height:25px;vertical-align: bottom;">' + v.type + '</span>'
+            );
+          });
+        });
```
Comment (linknum23) [inline web/static/js/stream-management.js:265]: This should be a function right? I see the similar code 3 different places

File: web/static/js/stream-management.js [hunk 4]
```diff
@@ -0,0 +1,488 @@
...
+    }
+
+    $.ajax({
+      type: "PATCH",
+      url: nurl + $("#edit-sid").val(),
+      data: JSON.stringify(formData),
+      contentType: "application/json",
+      success: function(data) {
+        $("#settings-tab-inputs-stream-title").text("Select a stream");
+        $("#settings-tab-inputs-stream-selection").empty();
```
Comment (linknum23) [inline web/static/js/stream-management.js:295]: here

File: web/static/js/stream-management.js [hunk 5]
```diff
@@ -0,0 +1,488 @@
...
+
+  /* Delete stream and reload stream list and settings */
+  $("#settings-tab-inputs-config").on("click", "#delete", function() {
+    $.ajax({
+      url: '/api/streams/' + document.getElementById("edit-sid").value,
+      type: 'DELETE',
+      success: function(data) {
+        // Reload stream list and settings
+        $("#settings-tab-inputs-stream-title").text("Select a stream");
+        $("#settings-tab-inputs-stream-selection").empty();
```
Comment (linknum23) [inline web/static/js/stream-management.js:331]: here

File: web/static/js/stream-management.js [hunk 6]
```diff
@@ -0,0 +1,488 @@
+/* On page load, get list of current streams */
+$(function() {
+  var streams = [];
+  $.get("/api/sources", function(data) {
```
Comment (linknum23) [inline web/static/js/stream-management.js:4]: here as well?

File: web/static/stream-management.html [hunk 7]
```diff
@@ -0,0 +1,85 @@
+<!DOCTYPE html>
```
Comment (linknum23) [inline web/static/stream-management.html:1]: This is all integrated into index.html.j2 right?

File: amplipi/app.py [hunk 8]
```diff
@@ -649,6 +649,13 @@ def doc():
   # TODO: add hosted python docs as well
   return FileResponse(f'{TEMPLATE_DIR}/rest-api-doc.html')
 
+# Stream Management
+
+@app.get('/manage', include_in_schema=False)
+def manage():
+  """ Get the Stream Management page """
+  return FileResponse(f'{STATIC_DIR}/stream-management.html')
+
```
Comment (linknum23) [inline amplipi/app.py:658]: Remove this? No need to have a separate page with the same functionality right?
```suggestion
```

**Review by linknum23 — CHANGES_REQUESTED** [review]

**Review by linknum23 — APPROVED** [review]
